src/Symfony/Component/HttpFoundation/Request.php

```diff
@@ -1220,7 +1229,12 @@ public function getHost()
         // check that it does not contain forbidden characters (see RFC 952 and RFC 2181)
         // use preg_replace() instead of preg_match() to prevent DoS attacks with long host names
         if ($host && '' !== preg_replace('/(?:^\[)?[a-zA-Z0-9-:\]_]+\.?/', '', $host)) {
-            throw new \UnexpectedValueException(sprintf('Invalid Host "%s"', $host));
+            if (!$this->isHostValid) {
```

Comment:
That's what I tried to explained in the description.

The first time you try to get the host (it happens in the request validation or router listener for instance), you get the exception, which means that you should then not access the request again. But, when handling the exception in the kernel, you might handle another request in development or production to convert it to a proper HTTP response (think API, ...). The problem comes from the fact that the request used to handle this exception is a duplicate of the original request. If we throw an exception again here, which will happen without this ugly code, then exception handling fails and the original exception bubbles. This is what we are trying to avoid here.